### PR TITLE
Client-supported features end-point

### PR DESCRIPTION
**Pull request by fisx**
https://wearezeta.atlassian.net/browse/SQSERVICES-417

changelog update (and possibly a few lines of documentation) coming up.

TODO:

- [x] change `legalhold` keyword to `legalhold-implicit-consent`
- [x] rename `feature` to `capability`

### CODE REVIEW

File: services/brig/src/Brig/API/Client.hs [hunk 1]
```diff
@@ -131,6 +137,11 @@ updateClient u c r = do
   unless ok $
     throwE ClientNotFound
   for_ (updateClientLabel r) $ lift . Data.updateClientLabel u c . Just
+  for_ (updateClientSupportedFeatures r) $ \features' -> do
+    features <- fromMaybe [] <$> Data.lookupClientSupportedFeatures u c
+    if Set.fromList features `Set.isSubsetOf` Set.fromList features'
+      then lift . Data.updateClientSupportedFeatures u c . Just $ features'
+      else throwE ClientFeaturesCannotBeRemoved
```
Comment (jschaul) [inline services/brig/src/Brig/API/Client.hs:144]: This doesn't make sense. Of course features can be removed, or changed, e.g. support for implicit consent could be removed, and instead support for explicit consent added. Why force clients in 3 years to support half-implemented features from the distant past?

File: services/brig/src/Brig/API/Public.hs [hunk 2]
```diff
@@ -818,6 +818,17 @@ sitemap o = do
     Doc.returns (Doc.ref Public.modelClient)
     Doc.response 200 "Client" Doc.end
 
+  get "/clients/:client/supported-features" (continue getClientSupportedFeaturesH) $
```
Comment (jschaul) [inline services/brig/src/Brig/API/Public.hs:821]: Is this endpoint passed through in the [nginz config](https://github.com/zinfra/backend-wiki/wiki/Checklists#adding-new-endpoints-in-wire-server) ?

File: services/brig/src/Brig/Data/Client.hs [hunk 3]
```diff
@@ -233,6 +244,9 @@ insertClient = "INSERT INTO clients (user, client, tstamp, type, label, class, c
 updateClientLabelQuery :: PrepQuery W (Maybe Text, UserId, ClientId) ()
 updateClientLabelQuery = "UPDATE clients SET label = ? WHERE user = ? AND client = ?"
 
+updateClientSupportedFeaturesQuery :: PrepQuery W (Maybe [SupportedClientFeature], UserId, ClientId) ()
```
Comment (jschaul) [inline services/brig/src/Brig/Data/Client.hs:247]: This should be a cassandra set (`C.Set` from  `ìmport Cassandra as C`) and not a Maybe List. See e.g. https://github.com/wireapp/wire-server/blob/develop/services/galley/src/Galley/Data/Queries.hs#L206-L207

File: services/brig/test/integration/API/User/Client.hs [hunk 4]
```diff
@@ -518,11 +529,38 @@ testUpdateClient opts brig = do
...
+              const (Just "client-feature-cannot-be-removed") === fmap Error.label . responseJsonMaybe
+
+        getClientSupportedFeatures brig uid (clientId c) !!! do
+          const 200 === statusCode
+          const (Just (SupportedClientFeatureList featuresOut)) === responseJsonMaybe
+
+  checkUpdate (Just [ClientSupportsLegalholdImplicitConsent]) True [ClientSupportsLegalholdImplicitConsent]
+  checkUpdate Nothing True [ClientSupportsLegalholdImplicitConsent]
+  checkUpdate (Just []) False [ClientSupportsLegalholdImplicitConsent]
+
```
Comment (jschaul) [inline services/brig/test/integration/API/User/Client.hs:563]: Can you test that updating the client with only the supportedClientFeature doesn't break the rest (i.e. other parts of the client entry are not erased) ?

File: docs/reference/cassandra-schema.cql [hunk 5]
```diff
@@ -1158,6 +1158,7 @@ CREATE TABLE brig_test.clients (
     lat double,
     lon double,
     model text,
+    supported_features list<int>,
```
Comment (jschaul) [inline docs/reference/cassandra-schema.cql:1161]: This should be a set, see comment further down.

**Review by jschaul — APPROVED** [review]
See inline comments for things to improve; the functionality of storing client input is implemented alright. 
However as I don't really see what the plan is to use this information afterwards, this is only a tentative approval of this particular storing functionality, not of a first part of a general plan. i.e. this code may get reverted entirely in the future if the plan how to make use of these features doesn't materialize.
For upcoming PRs please get reviews on the implementation ideas first before writing a lot of code.

File: libs/wire-api/src/Wire/API/User/Client.hs [hunk 6]
```diff
@@ -555,6 +565,54 @@ instance FromJSON UpdateClient where
       <$> o .:? "prekeys" .!= []
       <*> o .:? "lastkey"
       <*> o .:? "label"
+      <*> o .:? "supported-features"
+
+-- FUTUREWORK: add golden tests for this?
+data SupportedClientFeatureList = SupportedClientFeatureList {fromSupportedClientFeatureList :: [SupportedClientFeature]}
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform SupportedClientFeatureList)
+  deriving (ToSchema) via (CustomSwagger '[FieldLabelModifier (StripPrefix "fromSupportedClient", CamelToSnake)] SupportedClientFeatureList)
```
[the remark this answers is not in the document]
Comment (fisx) [reply libs/wire-api/src/Wire/API/User/Client.hs:586]: proved to be non-trivial, but i'll figure it out here, later: https://github.com/wireapp/wire-server/pull/1512

File: services/brig/src/Brig/API/Client.hs (hunk 1 above)
Comment (fisx) [reply services/brig/src/Brig/API/Client.hs:144]: It does make sense if we see this as a data-minimizing substitute for version information.  Versions usually aren't decreasing either, and I think it's reasonable to disallow it.

File: services/brig/src/Brig/API/Public.hs (hunk 2 above)
Comment (fisx) [reply services/brig/src/Brig/API/Public.hs:821]: Good point, thanks!  I'll add that.

File: services/brig/test/integration/API/User/Client.hs (hunk 4 above)
Comment (fisx) [reply services/brig/test/integration/API/User/Client.hs:563]: b9469683055876b3927c255cd1098a06202ebd8b

File: services/brig/schema/src/V64_ClientFeatures.hs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  )
+where
+
+import Cassandra.Schema
+import Imports
+import Text.RawString.QQ
+
+migration :: Migration
+migration = Migration 64 "Add supported client features to clients table" $ do
+  schema' [r| alter columnfamily clients add supported_features list<int>; |]
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/schema/src/V64_ClientFeatures.hs]: 10f1da8109780c034b41a7345f3fc380b4ffb8b9

File: services/brig/src/Brig/API/Client.hs (hunk 1 above)
Comment (fisx) [reply services/brig/src/Brig/API/Client.hs:144]: Another thing that can be said about in favor of disallowing it is that the restriction can be trivially lifted whenever we change our mind.

File: services/brig/src/Brig/API/Public.hs (hunk 2 above)
Comment (fisx) [reply services/brig/src/Brig/API/Public.hs:821]: actually, it looks like we're cool:

```yaml
    - path: /clients
      envs:
      - all
```

File: services/brig/src/Brig/API/Client.hs (hunk 1 above)
Comment (jschaul) [reply services/brig/src/Brig/API/Client.hs:144]: Okay, that's true, this check can be removed again.

File: services/brig/src/Brig/API/Public.hs (hunk 2 above)
Comment (jschaul) [reply services/brig/src/Brig/API/Public.hs:550]: Perfect!

**Review by jschaul — APPROVED** [review]

**Pull request merged**